python/rikai/spark/sql/codegen/mlflow_registry.py

```diff
@@ -0,0 +1,275 @@
+#  Copyright (c) 2021 Rikai Authors
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import os
+from typing import Any, Callable, Dict, IO, Mapping, Optional, Union
+from urllib.parse import urlparse
+
+try:
+    import mlflow
+except ImportError:
+    raise ImportError(
+        "Couldn't import mlflow. Please make sure to "
+        "`pip install mlflow` explicitly or install "
+        "the correct extras like `pip install rikai[mlflow]`"
+    )
+from mlflow.exceptions import MlflowException
+from mlflow.tracking import MlflowClient
+from pyspark.sql import SparkSession
+
+from rikai.logging import logger
+from rikai.spark.sql.codegen.base import (
+    ModelSpec,
+    register_udf,
+    Registry,
+    udf_from_spec,
+)
+from rikai.spark.sql.exceptions import SpecError
+
+__all__ = ["MlflowRegistry"]
+
+
+class MlflowModelSpec(ModelSpec):
+    """Model Spec.
+
+    Parameters
+    ----------
+    run : mlflow.entities.Run
+        the Run object containing the spec info and ref to model artifact
+    options : Dict[str, Any], optional
+        Additionally options. If the same option exists in spec already,
+        it will be overridden.
+    validate : bool, default True.
+        Validate the spec during construction. Default ``True``.
+    """
+
+    def __init__(
+        self,
+        run: "mlflow.entities.Run",
+        options: Optional[Dict[str, Any]] = None,
+        tracking_uri: str = None,
+        validate: bool = True,
+    ):
+        self._run = run
+        spec = self._run_to_spec_dict()
+        spec.setdefault("options", {})
+        if options:
+            spec["options"].update(options)
+        super().__init__(spec, validate=validate)
+        self._artifact = None
+        self.tracking_uri = tracking_uri
+
+    @property
+    def run_id(self):
+        return self._run.info.run_id
+
+    @property
+    def start_time(self):
+        return self._run.info.start_time
+
+    @property
+    def end_time(self):
+        return self._run.info.end_time
+
+    @property
+    def artifact(self) -> Any:
+        """Return Model artifact"""
+        if self._artifact is None:
+            mlflow.set_tracking_uri(self.tracking_uri)
+            self._artifact = getattr(mlflow, self.flavor).load_model(self.uri)
```

Comment:
does this mean that model will be picked via the content and send to workers?

I can see models that are memory bound (i.e., more than 10GB), will that be problematic? Also, this might means that the same model is in JVM, python interpreter as well as in GPU.

Why not call `load_model` in the `pandas_udf` itself tho? in that case, it does not take any spark/jvm seen memory.   Also this might be dangerous to load a bunch of models at once?

Reply:
yeah, potentially. Need to make sure to get the tracking_uri to the worker too is all.

Reply:
switched to load_model instead